### ISSUE TITLE
Can't work with company-mode

### ISSUE THREAD

**Issue by netjune** [joaotavora/yasnippet#996]
Here is a screen record video which shows the problem. http://121.43.160.118:3080/upload/yasnippet_bug2.mp4

Reproduce steps:
1. emacs -Q
2. install these packages from melpa:  yasnippet, company-mode, go-mode, company-go
3. install golang and package gocode: go get -u github.com/mdempsky/gocode
4. open the test file: a.go
5. set `company-backends` to '(company-go)
6. turn on company-mode and yas-minor-mode
7. do as in the screen record video

The test file a.go and the snippets used are uploaded to: http://121.43.160.118:3080/upload/yasnippets_bug2_files.tar.bz2

The order of snippets in the video:
1. sel
2. ch

This problem happens almost everyday for me. But it can't be 100% reproduced before until now.

**Comment by npostavs** [comment]
Thanks for the report! Actually it's not dependent on company or go-mode, here's a test case that shows the same failure:
```elisp
(ert-deftest nested-snippet-expansion-5-nested-delete ()
  "See Github #996."
  ;;:expected-result :failed
  (let ((yas-triggers-in-field t))
    (yas-with-snippet-dirs
     '((".emacs.d/snippets"
        ("text-mode"
         ("sel" . "${1:ch}")
         ("ch" . "<-${1:ch}"))))
     (yas-reload-all)
     (text-mode)
     (yas-minor-mode +1)
     (insert "sel")
     (ert-simulate-command '(yas-expand))
     (ert-simulate-command '(forward-word 1))
     (ert-simulate-command '(yas-expand))
     (ert-simulate-command '(forward-word 1))
     ;; Assertion failure here!
     (ert-simulate-command '(delete-backward-char 1))
     (should (string= (buffer-string) "<-c\n")))))
```

**Comment by netjune** [comment]
Thank for your confirmation.

**Comment by npostavs** [comment]
Should be fixed in #1002. https://raw.githubusercontent.com/npostavs/yasnippet/track-before-modifications/yasnippet.el

**Comment by netjune** [comment]
It works well with the latest company-mode now. Thanks.
